dsa-solutions/gfg-solutions/0005-next-happy-number.md

```diff
@@ -0,0 +1,255 @@
+---
+id: next-happy-number
+title: Next Happy Number (Geeks for Geeks)
+sidebar_label: 0004 - Next Happy Number
+tags:
+  - intermediate
+  - Fibonacci
+  - Dynamic Programming
+  - Mathematics
+  - Algorithms
+---
+
+This tutorial contains a complete walk-through of the Next Happy Number problem from the Geeks for Geeks website. It features the implementation of the solution code in three programming languages: Python, C++, and Java.
+
+## Problem Statement
+
+For a given non-negative integer N, find the next smallest Happy Number. A number is called Happy if it leads to 1 after a sequence of steps, where at each step the number is replaced by the sum of squares of its digits. If we start with a Happy Number and keep replacing it with the sum of squares of its digits, we eventually reach 1.
+
+## Examples
+
+**Example 1:**
+
+```
+input:
+N = 8
+
+Output:
+10
+```
+
+Explanation:
+Next happy number after 8 is 10 since
+$[ 1 * 1 + 0 * 0 = 1]$
+
+**Example 2**
+
+```
+Input:
+N = 10
+
+Output:
+13
+```
+
+Explanation:
+
+After 10, 13 is the smallest happy number because
+$[1 * 1 + 3 * 3 = 10]$, so we replace 13 by 10 and $[1 * 1 + 0 * 0 = 1]$.
+
+## Task
+
+You don't need to read input or print anything. Your task is to complete the function `nextHappy()` which takes an integer N as input parameters and returns an integer, the next Happy number after N.
+
+## Constraints
+
+1.  $(1 \leq N \leq 10^5)$
+
+## Solution Approach
+
+### Intuition
+
+To solve the problem, we need to:
+
+1. Identify the next number greater than N.
+2. Check if it is a Happy Number.
+3. Repeat the process until we find the next Happy Number.
+
+A number is identified as Happy if, by repeatedly replacing it with the sum of squares of its digits, we eventually reach 1.
+
+### Steps
+
+1. Implement a helper function to determine if a number is Happy.
+2. Start checking numbers greater than N, using the helper function to identify the next Happy Number.
+
+### Detailed Explanation
+
+The numbers that, when you repeatedly sum the squares of their digits, eventually result in 1 are known as "happy numbers."
+
+Here are examples of how to determine if numbers less than 10 are happy numbers:
+
+- **Number 1:**
+  $[1 ^ 2 = 1]$
+  Since we have already reached 1, the process stops here. 1 is a happy number.
+
+- **Number 2:**
+
```

Comment:
**Bad Practice:**

````md
```
  $[2^2 = 4]$
  $[4^2 = 16]$
  $[1^2 + 6^2 = 37]$
  $[3^2 + 7^2 = 58]$
  $[5^2 + 8^2 = 89]$
  $[8^2 + 9^2 = 145]$
  $[1^2 + 4^2 + 5^2 = 42]$
  $[4^2 + 2^2 = 20]$
  $[2^2 + 0^2 = 4]$
  Since we have reached 4 again, the process will continue in an infinite loop. 2 is not a happy number.
```
````

**Good Practice**

```md

 $[2^2 = 4]$
  $[4^2 = 16]$
  $[1^2 + 6^2 = 37]$
  $[3^2 + 7^2 = 58]$
  $[5^2 + 8^2 = 89]$
  $[8^2 + 9^2 = 145]$
  $[1^2 + 4^2 + 5^2 = 42]$
  $[4^2 + 2^2 = 20]$
  $[2^2 + 0^2 = 4]$

Since we have reached 4 again, the process will continue in an infinite loop. 2 is not a happy number.
```